### PR TITLE
ubuntu support

### PR DESCRIPTION
**Pull request by yuyang0**
pls review

### CODE REVIEW

File: playbooks/roles/calico/tasks/main.yaml [hunk 1]
```diff
@@ -7,14 +7,24 @@
 - name: ensure modules auto load
   copy: src=calico.conf dest=/etc/modules-load.d/calico.conf
 
+- include_vars: "{{ item }}"
```
Comment (bibaijin) [inline playbooks/roles/calico/tasks/main.yaml]: 是否可以使用 role 的依赖来实现变量导入？

File: playbooks/roles/dnsmasq/templates/dnsmasq.service.j2 [hunk 2]
```diff
@@ -0,0 +1,28 @@
+[Unit]
```
Comment (bibaijin) [inline playbooks/roles/dnsmasq/templates/dnsmasq.service.j2]: 是否可以用 lineinfile 来控制是否有某一行，而且指定系统为 ubuntu?

File: playbooks/roles/docker-version/tasks/main.yaml [hunk 3]
```diff
@@ -19,13 +19,20 @@
...
   yum: pkg={{ item }}
   with_items:
     - "/tmp/lain/docker-engine-selinux-{{ docker_version }}.noarch.rpm"
     - "/tmp/lain/docker-engine-{{ docker_version }}.x86_64.rpm"
   register: pkg
+  when: ansible_distribution == "CentOS"
+
+- name: install docker for ubuntu
+  apt:
+    deb: "/tmp/lain/docker-ce_17.09.1_ubuntu_{{ansible_distribution_version}}_amd64.deb"
```
Comment (bibaijin) [inline playbooks/roles/docker-version/tasks/main.yaml]: 17.09 是否可以使用变量？

File: playbooks/roles/docker/templates/docker.service.j2 [hunk 4]
```diff
@@ -0,0 +1,36 @@
+[Unit]
```
Comment (bibaijin) [inline playbooks/roles/docker/templates/docker.service.j2]: 是否可以使用单独修改某一行，同时判断系统版本？

File: playbooks/roles/etcd/files/etcd.conf [hunk 5]
```diff
@@ -0,0 +1,51 @@
+# [member]
```
Comment (bibaijin) [inline playbooks/roles/etcd/files/etcd.conf]: centos 下 etcd 的环境变量在 /var/etcd/env，这里需要斟酌~

File: playbooks/roles/etcd/templates/etcd.service.j2 [hunk 6]
```diff
@@ -1,9 +1,11 @@
 [Unit]
-Description=etcd
+Description=etcd key-value store
+Documentation=https://github.com/coreos/etcd
 After=network.target
 
 [Service]
 User=etcd
+Type=notify
```
Comment (bibaijin) [inline playbooks/roles/etcd/templates/etcd.service.j2]: 这里是什么作用？

File: playbooks/roles/firewall/tasks/main.yaml [hunk 7]
```diff
@@ -2,21 +2,35 @@
...
 - name: enable iptables services
   service: name=iptables enabled=yes
+  when: ansible_distribution == "CentOS"
 
 - name: deploy init iptables script
   template: src=iptables.sh.j2 dest=/tmp/iptables.sh
 
 - name: init iptables
-  shell: sh /tmp/iptables.sh
+  shell: bash /tmp/iptables.sh
```
Comment (bibaijin) [inline playbooks/roles/firewall/tasks/main.yaml]: sh 会出现什么问题？我看到 shebang 是 sh~

File: playbooks/roles/libraries/library/etcd_set_key.py [hunk 8]
```diff
@@ -30,16 +33,18 @@ def get_key(key):
         return data['node']['value']
 
     def set_key(key, value):
-        check_call(['etcdctl', 'set', key, value])
+        fnull = open(os.devnull, 'w')
```
Comment (bibaijin) [inline playbooks/roles/libraries/library/etcd_set_key.py]: fnull 是否需要 close，或者使用 with 语法？

File: playbooks/roles/rsync/templates/rsyncd.service.j2 [hunk 9]
```diff
@@ -0,0 +1,9 @@
+[Unit]
```
Comment (bibaijin) [inline playbooks/roles/rsync/templates/rsyncd.service.j2]: 这个 service 没问题，与 centos 兼容~

File: playbooks/roles/calico/tasks/main.yaml (hunk 1 above)
Comment (yuyang0) [reply playbooks/roles/calico/tasks/main.yaml]: good idea

File: playbooks/roles/etcd/templates/etcd.service.j2 (hunk 6 above)
Comment (yuyang0) [reply playbooks/roles/etcd/templates/etcd.service.j2]: systemd log warnning message when without Type, you can see (here)[https://github.com/coreos/etcd/blob/master/contrib/systemd/etcd.service]

File: playbooks/roles/firewall/tasks/main.yaml (hunk 7 above)
Comment (yuyang0) [reply playbooks/roles/firewall/tasks/main.yaml]: ubuntu's `sh` is symbol link to dash, it seems not support the function definition style in iptables.sh

File: playbooks/roles/docker-version/tasks/main.yaml (hunk 3 above)
Comment (yuyang0) [reply playbooks/roles/docker-version/tasks/main.yaml]: if rpm's version and deb's version are same, we can replace it with `docker_ver`, but rpm's version is too old

File: playbooks/roles/libraries/library/etcd_set_key.py (hunk 8 above)
Comment (yuyang0) [reply playbooks/roles/libraries/library/etcd_set_key.py]: yeah, it's better to close it explicitly, but it is also ok if not close here, when plugin's process exits, open files will be closed automatically.

**Comment by yuyang0** [comment]
everything seems ok, please review again.

File: playbooks/roles/docker-version/tasks/main.yaml (hunk 3 above)
Comment (bibaijin) [reply playbooks/roles/docker-version/tasks/main.yaml]: OK, we will upgrade docker in centos later, and then define `docker_ver = 17.09` .

File: playbooks/roles/etcd/templates/etcd.service.j2 (hunk 6 above)
Comment (bibaijin) [reply playbooks/roles/etcd/templates/etcd.service.j2]: OK, it looks good to me.

File: playbooks/roles/firewall/tasks/main.yaml (hunk 7 above)
Comment (bibaijin) [reply playbooks/roles/firewall/tasks/main.yaml]: LGTM

**Comment by bibaijin** [comment]
@yuyang0 LGTM.

**Pull request merged**